feat: replace amqplib with amqp-client.js

semurgx opened the pull request.

alexkvak commented:
All threads from #45 should be answered

alexkvak commented:
Overall coverage decreased (-2.6%) to 86.707%

semurgx commented:
Increased coverage and extended commit message.

alexkvak commented:
What is the necessity in these breaking changes?

 It's easy to preserve old interface at least in case of Date/number type — convert timestamp into Date in `ServiceConnection.postMessage`

alexkvak commented on src/message.ts:
```diff
@@ -1,5 +1,5 @@
-interface MessageHeaders {
-  recipients: string;
+type MessageHeaders = {
```
This is kind of holy-war change, why you did it?

semurgx replied:
I bought back interface here.

alexkvak commented on src/message.ts:
```diff
@@ -59,7 +57,7 @@ interface DistributedMessageProperties {
    * won't be mutated
    */
-  headers: MessageHeaders;
+  headers?: MessageHeaders;
```
Why the Message interface changed? If it is necessary for new library the commit message should contain detailed information about this.

semurgx replied:
I've brought library's old interface.

alexkvak commented on tsconfig.json:
```diff
@@ -3,7 +3,7 @@
     "module": "commonjs",
     "target": "es2018",
-    "lib": ["es6", "es2017"],
+    "lib": ["dom", "es6", "es2017"],
```
why the dom lib is needed?

semurgx replied:
Left by mistake.

alexkvak commented on src/adapters/amqp-node.ts:
```diff
@@ -24,61 +20,33 @@ export interface QueueOptions {
-  cancel: (consumerTag: string) => Promise<void>;
-  close: () => Promise<void>;
-}
+export type AMQPConnection = Pick<AMQPChannel, 'basicAck' | 'basicCancel' | 'basicConsume' | 'basicNack' | 'basicPublish' | 'close' | 'exchangeDeclare' | 'prefetch' | 'queue' | 'queueBind'>;
```
It is safer to declare own interfaces in mbclient rather than blindly reexport from amqp-client.

That way allows us explicitly control types and handle their mismatch immediately.

semurgx replied:
Reexported only used amp-client.js type subset.

alexkvak commented on src/service.spec.ts:
```diff
@@ -198,39 +192,19 @@ describe('#handleConnectionClose', () => {
-describe('#connectionEventHandler', () => {
-  const message = { content: 'message' };
-
-  it('calls connection error handler on error event', () => {
```
Does connectionEventHandler not call connection error handler on error event for now?

semurgx replied:
amqplib supported multiple connection and channel events which from actually handled only `close` event. Everthing else just logged. amqp-client.js supports only `close` event and I decided that dispatcher method `connectionEventHandler` is redundant. `handlerConnectionClose` is directly called in `onerror` callback,

semurgx commented:
I brought back old interfaces.

semurgx commented:
Final commit messages:
feat: replace amqplib with amqp-client.js

Added message body which contains raw data as Uint8Array.
Removed unused event handlers on connection and channel.

alexkvak commented on src/adapters/amqp-node.ts:
```diff
@@ -17,68 +12,77 @@ export interface AMQPOptions {
+interface AMQPConsumer { tag: string; }
+interface AMQPExchangeParams { durable: true; }
+type AMQPExchangeType = 'topic';
+export type AMQPMessageCallback = (message: AMQPMessage) => void;
```
It’s better to declare types in the same file with their usage

alexkvak commented:
:tada: This PR is included in version 3.1.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/mbclient/v/3.1.0)
- [GitHub release](https://github.com/Tinkoff/mbclient/releases/tag/v3.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
